components/raftstore/src/store/peer.rs

```diff
@@ -1974,12 +1975,26 @@ where
     }
 
     /// Responses to the ready read index request on the replica, the replica is not a leader.
-    fn post_pending_read_index_on_replica<T>(&mut self, ctx: &mut PollContext<EK, ER, T>) {
+    fn post_pending_read_index_on_replica<T: Transport>(
+        &mut self,
+        ctx: &mut PollContext<EK, ER, T>,
+    ) {
         while let Some(mut read) = self.pending_reads.pop_front() {
-            // addition_request indicates an ongoing lock checking. We must wait until lock checking finished.
-            if read.addition_request.is_some() {
-                self.pending_reads.push_front(read);
-                break;
+            // The response of this read index request is lost, but we need it for
+            // the memory lock checking result. Re-propose the request.
+            if let Some(read_index) = read.addition_request.take() {
+                assert_eq!(read.cmds.len(), 1);
+                let (mut req, cb, _) = read.cmds.pop().unwrap();
+                assert_eq!(req.requests.len(), 1);
+                req.requests[0].set_read_index(*read_index);
+                info!(
+                    "re-propose read index request because the response is lost";
+                    "region_id" => self.region_id,
+                    "peer_id" => self.peer.get_id(),
+                );
+                RAFT_READ_INDEX_PENDING_COUNT.sub(1);
+                self.propose(ctx, cb, req);
```

Comment:
After introducing async ready from https://github.com/tikv/tikv/pull/9177, we can not propose commands while there is a pending `Ready` not `advance_append` yet, maybe we should send a `PeerMsg::RaftCommand` instead. /cc @gengliqi

Reply:
IMO `propose` here only proposes a read index request which doesn't depend on the current state of raft (it will check the epoch version and other states before handling the read request later).
Not sure if there is any case that breaks the correctness.

Reply:
Although I can't remember the sepecified case, but we come to a conclusion that `raft::step` should not be called when there is a pending `Ready` not `advance_append` yet, PTAL @gengliqi, also propose read index request may cause propose an extra empty command is some cases:
https://github.com/tikv/tikv/blob/47e03ae7906e7c32208ea01521416bae6c66ad30/components/raftstore/src/store/peer.rs#L2694-L2707

Reply:
I choose to copy the unstable entries to `Ready` in async ready so maybe it's ok to propose `read_index` when there is a pending `Ready`.
However, there is another problem here. `post_pending_read_index_on_replica` may be called after getting the `Ready` which means the msg can not be sent until the next `Ready`.
In the previous logic, the `read_index` requests retry in `on_raft_base_tick`. It's ok because the `Ready` will be got later.